### PR TITLE
Delaunay3D/LightmapGI: Improve triangulation

### LINKED ISSUE #83972: LightmapGi bake freeze when generating volume probes for a large level with large amount of probes
### Godot version

v4.2.beta.custom_build [e8d57afae]

### System information

Ubuntu with Nvidia 2060

### Issue description

When baking probes for the TPS demo with a lot of manual LightProbes placed across the level, the baking process got stuck forever without doing anything in `Generating Light Probes Volume`

### Steps to reproduce

Download MRP (it is the TPS demo setted with lightmap probes ready to bake)
Open Level2.tsn
Bake the lightmap

### Minimal reproduction project

https://www.dropbox.com/scl/fi/vvz4yeoa4b4aaqmu4pjjf/tps-demo-4-test.zip?rlkey=h5qidqjqcgg62ulx0thsecxhl&dl=0

### LINKED ISSUE #90138: Delaunay3D can produce overlapping tetrahedra
### Tested versions

4.3.dev.custom_build.29b3d9e9e

### System information

Godot v4.3.dev (248e58a4e) - Debian GNU/Linux trixie/sid trixie - X11 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2080 (nvidia) - Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz (8 Threads)

### Issue description

`Delaunay3D::tetrahedralize()` when called directly from C++ or from GDScript via `Geometry3D.tetrahedralize_delaunay()` can produce overlapping tetrahedra (with points inside their circumsphere). All examples of two overlapping tetrahedra that I've seen have had two or three vertices in common and intersecting volumes.

screenshot

It doesn't happen all the time and I don't know exactly what circumstances trigger it, but it seems that both of these must be true:
- there are some points that are several hundreds distance units apart, and
- there are some points that are only a couple of units apart

I noticed this problem while working on issues with LightmapGI probe triangulation and partitioning.

It's difficult to create an MRP that doesn't involve thousands of points and doesn't require extra tooling to find and visualize the few invalid cases. The attached MRP is the smallest and simplest thing I can do.

The MRP is just one scene with one node and one script. I'll dump the script here so you can check my code for obvious bugs without running the project.

```gdscript
@tool
extends MeshInstance3D

const points: PackedVector3Array = [
	Vector3(204.537, 2.64082, 0),
	Vector3(208.53, 2.64082, 0),
	Vector3(204.537, 2.64082, 2),
	Vector3(206.53, 2.64082, 2),
	Vector3(-501.4156, -2.415572, -501.4156),
	Vector3(501.4156, -2.415572, -501.4156),
	Vector3(-501.4156, 248.2922, -501.4156),
	Vector3(501.4156, 248.2922, -501.4156),
	Vector3(-501.4156, -2.415572, 501.4156),
	Vector3(501.4156, -2.415572, 501.4156),
	Vector3(-501.4156, 248.2922, 501.4156),
	Vector3(501.4156, 248.2922, 501.4156),
	Vector3(-250.7078, -2.415572, -501.4156),
	Vector3(-250.7078, 248.2922, -501.4156),
	Vector3(-501.4156, -2.415572, -250.7078),
	Vector3(-250.7078, -2.415572, -250.7078),
	Vector3(-501.4156, 248.2922, -250.7078),
	Vector3(-250.7078, 248.2922, -250.7078),
	Vector3(0, -2.415572, -501.4156),
	Vector3(0, 248.2922, -501.4156),
	Vector3(0, -2.415572, -250.7078),
	Vector3(0, 248.2922, -250.7078),
	Vector3(-501.4156, -2.415572, 0),
	Vector3(-250.7078, -2.415572, 0),
	Vector3(-501.4156, 248.2922, 0),
	Vector3(-250.7078, 248.2922, 0),
	Vector3(0, -2.415572, 0),
	Vector3(0, 248.2922, 0),
	Vector3(250.7078, -2.415572, -501.4156),
	Vector3(250.7078, 248.2922, -501.4156),
	Vector3(250.7078, -2.415572, -250.7078),
	Vector3(250.7078, 248.2922, -250.7078),
	Vector3(501.4156, -2.415572, -250.7078),
	Vector3(501.4156, 248.2922, -250.7078),
	Vector3(250.7078, -2.415572, 0),
	Vector3(250.7078, 248.2922, 0),
	Vector3(501.4156, -2.415572, 0),
	Vector3(501.4156, 248.2922, 0),
	Vector3(-501.4156, -2.415572, 250.7078),
	Vector3(-250.7078, -2.415572, 250.7078),
	Vector3(-501.4156, 248.2922, 250.7078),
	Vector3(-250.7078, 248.2922, 250.7078),
	Vector3(0, -2.415572, 250.7078),
	Vector3(0, 248.2922, 250.7078),
	Vector3(-250.7078, -2.415572, 501.4156),
	Vector3(-250.7078, 248.2922, 501.4156),
	Vector3(0, -2.415572, 501.4156),
	Vector3(0, 248.2922, 501.4156),
	Vector3(250.7078, -2.415572, 250.7078),
	Vector3(250.7078, 248.2922, 250.7078),
	Vector3(501.4156, -2.415572, 250.7078),
	Vector3(501.4156, 248.2922, 250.7078),
	Vector3(250.7078, -2.415572, 501.4156),
	Vector3(250.7078, 248.2922, 501.4156)]

var imesh: ImmediateMesh

func _ready() -> void:
	imesh = ImmediateMesh.new()
	mesh = imesh

	var tetrahedra: PackedInt32Array = Geometry3D.tetrahedralize_delaunay(points)
	assert(tetrahedra.size() % 4 == 0)

	# These two are just one example of overlapping tetrahedra
	draw_tetrahedron(tetrahedra, 129, Color.RED)
	draw_tetrahedron(tetrahedra, 130, Color.BLUE)

func draw_tetrahedron(tetrahedra: PackedInt32Array, index: int, color: Color) -> void:
	var vertices: Array[Vector3] = [
		points[tetrahedra[index * 4]],
		points[tetrahedra[index * 4 + 1]],
		points[tetrahedra[index * 4 + 2]],
		points[tetrahedra[index * 4 + 3]]
	]
	print(vertices)

	imesh.surface_begin(Mesh.PRIMITIVE_TRIANGLES)
	for i in 4:
		imesh.surface_set_color(color)
		imesh.surface_add_vertex(vertices[i])
		imesh.surface_add_vertex(vertices[(i + 1) % 4])
		imesh.surface_add_vertex(vertices[(i + 2) % 4])
	imesh.surface_end()
```

The two tetrahedra chosen here, and in the screenshot above, are:
`[(250.7078, 248.2922, 0), (501.4156, 248.2922, 0), (250.7078, -2.415572, 250.7078), (250.7078, 248.2922, 250.7078)]`
and:
`[(250.7078, 248.2922, 0), (501.4156, -2.415572, 0), (250.7078, -2.415572, 250.7078), (250.7078, 248.2922, 250.7078)]`

I just noticed that in this particular example, the two have _exactly_ the same circumsphere, but that is likely a red herring since it is not true for many other cases (e.g. 117 together with 167). 

### Steps to reproduce

Load the MRP. It's a tool script and will show the issue directly in the editor.

### Minimal reproduction project (MRP)

[invalid-delaunay3d-mrp.zip](https://github.com/godotengine/godot/files/14837582/invalid-delaunay3d-mrp.zip)

### PR DESCRIPTION
**Pull request by permelin**
Fixes #90138
Fixes #83972

## A note about scope

The Delaunay3D class is only used for LightmapGI probes internally, but it is also exposed to GDScript. My focus here is the internal use for probe triangulation.

## The issues

1. Overlapping tetrahedra.
2. The pruning of thin/small tetrahedra removes too much.
3. Tetrahedra are stretched out along one or two axes.

## Overlapping tetrahedra

In a scene with a large enough volume and probes that are place just a few meters apart, triangulation can break down and generate overlapping tetrahedra.

I struggled to find a fix for this until @lawnjelly recommended quantization. I was skeptical, since quantization can flip a triangle around or obscure the fact that some points are coplanar. That would make things worse, not better. But I was also desperate, so I tried it. 

After some tuning I found that with enough bits, it actually did help.

Even though I'm no longer able to provoke invalid triangulation, I don't expect the triangulation to be bullet proof. I'm sure there are still configurations of points that will break it.

## Pruned tetrahedra

Once the triangulation is completed, a final pass discards any tetrahedra that are too thin or too small in terms of volume.

Having a fixed cut-off for volume is problematic since points are scaled down to the unit cube. In a 1000x1000 scene, tetrahedra that are several meters thick get pruned. This results in orphan probes that are not connected, and therefore not used. And after a lot of testing, I'm fairly convinced that this pruning serves no purpose. So have removed it, and instead made the thickness based pruning a bit more aggressive.

## Stretched triangulation

If the point cloud dimensions are oblong rather than cube shaped, triangulation is proportionally stretched out.

The left shows the current triangulation of random points in a 1000x1000x50 volume. The right is after scaling has been corrected.

delaunay-axis-scaling

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally (with #90701 applied on top), it works as expected.

This testing project previously resulted in 200+ warnings being printed on every bake, now it doesn't result in any warning: [test_lightmap_preview_bake_4.x.zip](https://github.com/godotengine/godot/files/14981644/test_lightmap_preview_bake_4.x.zip)[^1]

Code looks good to me at a glance. Thanks for commenting it thoroughly :slightly_smiling_face: 

[^1]: I intentionally placed multiple probes at the same location in the testing project to see if it breaks, but it appears to bake just fine and dynamic object lighting still behaves correctly.

File: core/math/delaunay_3d.h [hunk 1]
```diff
@@ -46,7 +46,8 @@ class Delaunay3D {
 	struct Simplex;
 
 	enum {
-		ACCEL_GRID_SIZE = 16
+		ACCEL_GRID_SIZE = 16,
+		QUANTIZATION_MAX = 1 << 16 // A power of two smaller than the 23 bit significand of a float
```
Comment (AThousandShips) [inline core/math/delaunay_3d.h:50]:
```suggestion
		QUANTIZATION_MAX = 1 << 16 // A power of two smaller than the 23 bit significand of a float.
```

**Comment by lawnjelly** [comment]
I haven't looked at this in detail as I'm not familiar, but can you not store the points as e.g. `Vector3i` (or 16 bit version) and take them out of the float domain completely?

**Comment by permelin** [comment]
You probably could keep the points as integers. I tried it briefly at first but had some issue I couldn't explain.

Since dividing a float with a power of two only moves the binary point to the left by changing the exponent, the quantization is not lost by doing so. The subsequent math done with the points couldn't possibly be kept in the integer domain anyway. And some of the intermediate values when calculating the circumsphere is very, very large. So I thought that I might as well take full advantage of the R128 fixed point instead of using just half of it.

**Comment by lawnjelly** [comment]
It is just a suggestion, as I say I haven't examined in detail so take with a pinch of salt.

Keeping as integer / fixed point as much as possible throughout can at least *potentially* exclude the possibly of more float bugs in other areas. I noticed e.g. you are using an epsilon in `simplex_is_coplanar()` and I'm wondering whether this is needed in integer space for instance. Likewise in the BSP this could potentially avoid bugs if epsilon is used (you can maybe do which side plane checks in integer math).

**Review by lawnjelly — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**